### DISCUSSION TITLE
How to place file in Android's external storage when installing application.

### DISCUSSION THREAD

**Discussion by Audionysos** [unoplatform/uno#7011] [Q&A] [answered]
## What would you like to be added:
I can't figure out how to place asset files in private external storage on Android.
From what I understand the the assets are embedded somewhere inside the executable/library,
Finally I thought to maybe copy them manually but I can't even access them using `Assets.Open()` as described in "AboutAssets.txt".
I have following structure in shared project:
```Assets/Audio/default/In The Mood.mp3```
I've set the **BuildAction** to **Content**. On Windows(UWP) the file is placed as expected, in the path relative to executable, and the program works fine.
On android I have no idea where the file is. I've printed all subdirectories from local storage as well as private external storage but I don't see it. So I've tried to open it using example form "AboutAssets.txt"
```
protected override void OnStart() {
	base.OnStart();

	var input = Assets.Open(@"Audio/default/In The Mood.mp3");
	WriteLine(input.Length);
	input.Close();
}
```
But the open method says the file is not found.
I can't find any info about the assets except at [Assets and image display](https://platform.uno/docs/articles/features/working-with-assets.html) but it doesn't help.

It would be nice if there was an example of how to deploy the files in external storage when application is installed or at least how can I access any asset file so I could copy it manually.

## Why is this needed:
I want the user to be able to place he's own sounds packages easily inside the common directory so it can be chosen instead of default one. For example user will create directories structure in `Audio\My Sounds\...` which reflects the original structure and then he can simply select "My Sounds" pack in the application's settings.

## For which Platform:

- [ ] iOS
- [x] Android
- [ ] WebAssembly
- [ ] WebAssembly renderers for Xamarin.Forms
- [ ] macOS
- [ ] Skia
  - [ ] WPF
  - [ ] GTK (Linux)
  - [ ] Tizen
- [ ] Windows

**Comment by robloo** [comment]
You need to add the file to project as an embedded resource. BuildAction -> EmbeddedResource. To use the data stream in a cross-platform way you can do something like:

```csharp
    Assembly assembly = Assembly.GetExecutingAssembly();
    string resourcePath = assembly.GetManifestResourceNames().FirstOrDefault(
        resourceName => resourceName.EndsWith("fileName" + "extension", StringComparison.OrdinalIgnoreCase));

    using (Stream stream = assembly.GetManifestResourceStream(resourcePath))
    {
        using (StreamReader reader = new StreamReader(stream))
        {
            var text = reader.ReadToEnd();
        }
    }
```

Copying the stream to an external location can be done using other tutorials for either mono/.NET or WinRT. You also need to make sure permissions are setup.

**Comment by Audionysos** [comment]
Ok, if that's the only way I will try it. I've already spend half of yesterday trying to figure out Android's file system and I think the copy thing should not be an issue, but setting file as embedder resource will be problematic. Apart form wasting storage space, the program was already working on windows but when I set the file to be embedded resource it would no longer appear in programs directory and root directories will be completely different...

I still don't understand where the file goes on android when i set it as "Content" in a shared project. The other .png file I use in the project is set to content and is reused for all platforms. Are "Content" assets limited to picture files only?

**Comment by robloo** [comment]
> as embedder resource will be problematic. Apart form wasting storage space, the program was already working on windows but when I set the file to be embedded resource it would no longer appear in programs directory

Oh, I see. I believe Uno should automatically transform Content into an Android Asset. I don't think Embeded Resource is the only way, but it is what I've used in the past. I don't need the file to appear in the file system with the executable though.

https://stackoverflow.com/a/64483046

<blockquote><img src="https://cdn.sstatic.net/Sites/stackoverflow/Img/apple-touch-icon@2.png?v=73d79a89bded" width="48" align="right"><div>Stack Overflow</div><div><strong><a href="https://stackoverflow.com/questions/64455707/content-build-action-for-resource-files-in-the-uno-platform">Content Build Action for Resource Files in the Uno Platform</a></strong></div><div>I have warnings building an Uno Platform app. Summarized, the warnings appear as below:
...\Source\Resources\Icon-256.svg : warning XA0101: @(Content) build action is not supported
...\Source\Prope...</div></blockquote>

**Comment by jeromelaban** [comment, marked as answer]
Check for [this document](https://platform.uno/docs/articles/features/windows-storage.html?q=GetFileFromApplicationUriAsync#support-for-storagefilegetfilefromapplicationuriasync) about `StorageFile.GetFileFromApplicationUriAsync()`, you can also take a look at [this sample](https://github.com/unoplatform/Uno.Samples/tree/master/UI/PackageResources).

**Comment by Audionysos** [comment]
Thank you guys, I will post how I did it when I succeed.
